# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


# Proyecto SQL

El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. Eso atrajo la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

Te han dado una base de datos de uno de los servicios que compiten en este mercado. Contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros. Esta información se utilizará para generar una propuesta de valor para un nuevo producto.

# Importar librerías

In [1]:
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})



In [2]:
query = "SELECT * FROM books LIMIT 5;"

In [3]:
pd.io.sql.read_sql(query, con = engine)

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


In [4]:
df = pd.read_sql(query, con=engine)

In [5]:
print(df.head())

   book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  


In [20]:
preview_queries = {
    "books":      "SELECT * FROM books LIMIT 5;",
    "authors":    "SELECT * FROM authors LIMIT 5;",
    "publishers": "SELECT * FROM publishers LIMIT 5;",
    "ratings":    "SELECT * FROM ratings LIMIT 5;",
    "reviews":    "SELECT * FROM reviews LIMIT 5;"
}

In [21]:
for name, q in preview_queries.items():
    print(f"\n### {name.upper()}")
    dfpre = pd.read_sql(q, con=engine)
    print(dfpre)


### BOOKS
   book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  

### AUTHORS
   author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5      

<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a><br> <b>Éxito</b> - Muy bien, lograste conectarte a la base de datos y mostrar una vista previa clara de todas las tablas. La estructura de datos está correctamente explorada y lista para análisis posteriores. </div>


# OBJETIVOS

1) Contar libros publicados después del **01-01-2000**.  
2) Para cada libro, obtener **nº de reseñas de texto** y **calificación promedio**.  
3) Hallar la **editorial** con más libros de **> 50 páginas**.  
4) Encontrar el **autor** con mayor **calificación promedio** considerando libros con **≥ 50 calificaciones**.  
5) Calcular el **promedio de reseñas de texto** entre usuarios que **calificaron > 50** libros.

# Encuentra el número de libros publicados después del 1 de enero de 2000.

In [6]:
query = """
SELECT COUNT(*) AS books_after_2000
FROM books
WHERE publication_date > DATE '2000-01-01';
"""

In [7]:
df = pd.read_sql(query, con=engine)
print(df)

   books_after_2000
0               819


son 819 libros publicados

<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a><br> <b>Éxito</b> - La consulta está bien planteada y el resultado es claro: lograste identificar correctamente los 819 libros publicados después de 2000. Muy buen inicio de análisis. </div>


# Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro. 

In [8]:
query = """
SELECT
    b.book_id,
    b.title,
    COALESCE(rv.num_reviews, 0) AS num_reviews,
    rt.avg_rating
FROM books b
LEFT JOIN (
    SELECT book_id, COUNT(*) AS num_reviews
    FROM reviews
    GROUP BY book_id
) rv USING (book_id)
LEFT JOIN (
    SELECT book_id, ROUND(AVG(rating), 2) AS avg_rating
    FROM ratings
    GROUP BY book_id
) rt USING (book_id)
ORDER BY num_reviews DESC NULLS LAST, avg_rating DESC NULLS LAST;
"""


In [9]:
df_books_reviews = pd.read_sql(query, con=engine)

In [10]:
df_books_reviews.head(15)

,book_id,title,num_reviews,avg_rating
0,948,Twilight (Twilight #1),7,3.66
1,302,Harry Potter and the Prisoner of Azkaban (Harr...,6,4.41
2,299,Harry Potter and the Chamber of Secrets (Harry...,6,4.29
3,656,The Book Thief,6,4.26
4,734,The Glass Castle,6,4.21
5,497,Outlander (Outlander #1),6,4.13
6,750,The Hobbit or There and Back Again,6,4.13
7,695,The Curious Incident of the Dog in the Night-Time,6,4.08
8,779,The Lightning Thief (Percy Jackson and the Oly...,6,4.08
9,963,Water for Elephants,6,3.98


<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a><br> <b>Éxito</b> - Excelente consulta, lograste combinar reseñas y calificaciones mostrando resultados claros y ordenados por relevancia. La salida refleja con precisión los libros más comentados junto a sus promedios de calificación. </div>


# Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis).

In [11]:
q_top1 = """
SELECT
    p.publisher,
    COUNT(*) AS books_over_50_pages
FROM publishers p
JOIN books b ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY books_over_50_pages DESC
LIMIT 1;
"""

In [12]:
edi_top1 = pd.read_sql(q_top1, con=engine)

In [13]:
edi_top1

,publisher,books_over_50_pages
0,Penguin Books,42


la editorial con mayor libros publicados es Penguin Books con 42

<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a><br> <b>Éxito</b> - Muy bien planteada la consulta, identificaste correctamente a la editorial con mayor número de publicaciones relevantes: *Penguin Books* con 42 libros. Resultado claro y útil para el análisis. </div>


# Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [14]:
query = """
WITH book_stats AS (
    SELECT
        b.book_id,
        b.author_id,
        COUNT(r.rating_id) AS n_ratings,
        AVG(r.rating)      AS avg_rating
    FROM books b
    JOIN ratings r ON r.book_id = b.book_id
    GROUP BY b.book_id, b.author_id
),
qualified AS (
    SELECT * 
    FROM book_stats
    WHERE n_ratings >= 50
)
SELECT
    a.author,
    ROUND(SUM(q.avg_rating * q.n_ratings) / SUM(q.n_ratings), 3) AS author_weighted_avg_rating,
    SUM(q.n_ratings) AS total_ratings_considered,
    COUNT(*)         AS qualifying_books
FROM qualified q
JOIN authors a ON a.author_id = q.author_id
GROUP BY a.author
ORDER BY author_weighted_avg_rating DESC
LIMIT 1;
"""

In [15]:
best_author = pd.read_sql(query, con=engine)

In [16]:
best_author

,author,author_weighted_avg_rating,total_ratings_considered,qualifying_books
0,J.K. Rowling/Mary GrandPré,4.287,310.0,4


La autora mas valorada y mejor calificada es JK Rowling autora de la saga Harry potter

<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a><br> <b>Éxito</b> - Excelente análisis, aplicaste bien la condición de mínimo de calificaciones y obtuviste un resultado sólido: J.K. Rowling es la autora mejor valorada con un promedio ponderado de 4.287. </div>


# Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [17]:
query = """
WITH raters AS (
    SELECT username, COUNT(*) AS ratings_count
    FROM ratings
    GROUP BY username
    HAVING COUNT(*) > 50
),
user_reviews AS (
    SELECT username, COUNT(*) AS reviews_count
    FROM reviews
    GROUP BY username
)
SELECT ROUND(AVG(COALESCE(ur.reviews_count, 0)), 2) AS avg_text_reviews_among_power_raters
FROM raters r
LEFT JOIN user_reviews ur USING (username);
"""

In [18]:
df = pd.read_sql(query, con=engine)

In [19]:
df

,avg_text_reviews_among_power_raters
0,24.33


Hay un promedio de 24.33 reseñas que hacen usuarios que dieron a mas de 50 libros su calificacion.

<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a><br> <b>Éxito</b> - Muy buen trabajo, la consulta está bien estructurada y el resultado es claro: los usuarios que calificaron más de 50 libros escriben en promedio 24.33 reseñas de texto. </div>


## Comentario general del revisor 

# Comentario General del Revisor

<div class="alert alert-block alert-success"> <b>Comentario del revisor</b> <a class="tocSkip"></a>  
    
¡Felicidades! Tu proyecto está **aprobado**. Lograste dar cumplimiento a todos los puntos solicitados en el brief: exploración de datos, consultas SQL específicas, obtención de resultados y conclusiones claras.  

#### Puntos Positivos:

* **Procesamiento de datos:** Conexión a la base de datos y extracción de información realizadas de forma correcta y ordenada.
* **Visualización y análisis gráfico:** La vista previa de tablas y resultados permite interpretar con facilidad la información obtenida.
* **Ingeniería de características:** Aplicaste filtros y condiciones relevantes (como más de 50 calificaciones) que enriquecieron el análisis.
* **Comparación de modelos:** Aunque no había modelado, lograste contrastar resultados entre autores, editoriales y usuarios, cumpliendo el espíritu de comparación requerido.

El proyecto demuestra un dominio adecuado de SQL y un análisis consistente, con resultados bien interpretados que fortalecen la propuesta de valor planteada. Excelente desempeño. </div>
